src/Model/HasAttributesTrait.php

```diff
@@ -231,6 +195,57 @@ public function toArray(): array
         return $this->getDocumentAttributes();
     }
 
+    /**
+     * {@inheritdoc}
+     */
+    public static function fill(
+        array $input,
+        ?HasAttributesInterface $object = null,
+        bool $force = false
+    ): HasAttributesInterface {
+        if (!$object) {
+            $object = Container::make(static::class);
+        }
+
+        if ($object instanceof PolymorphableModelInterface) {
+            $class = $object->polymorph(
+                array_merge($object->getDocumentAttributes(), $input)
+            );
+
+            if ($class !== $object::class) {
+                $originalAttributes = $object->getDocumentAttributes();
+                $object = new $class();
+
+                foreach ($originalAttributes as $key => $value) {
+                    $object->setDocumentAttribute($key, $value);
+                }
+            }
+        }
+
+        foreach ($input as $key => $value) {
+            if (
+                $force ||
+                (
+                    (!$object->fillable || in_array($key, $object->fillable)) &&
+                    !in_array($key, $object->guarded)
+                )
+            ) {
+                if ($value instanceof stdClass) {
+                    $value = json_decode(
+                        json_encode($value, JSON_THROW_ON_ERROR),
+                        true,
+                        512,
+                        JSON_THROW_ON_ERROR
+                    ); // cast to array
+                }
+
+                $object->setDocumentAttribute($key, $value);
+            }
```

Comment:
The new structure of the `if` it's better, but I think we can do even better if we avoid use `if` inside of `if`.

```suggestion
            if (
                !$force && 
                (
                    ($object->fillable && !in_array($key, $object->fillable)) || 
                    in_array($key, $object->guarded)
                )
            ) {
                continue;
            }
            
            if ($value instanceof stdClass) {
                $value = json_decode(
                    json_encode($value, JSON_THROW_ON_ERROR),
                    true,
                    512,
                    JSON_THROW_ON_ERROR
                ); // cast to array
            }

            $object->setDocumentAttribute($key, $value);
```

You should test this logic, but maybe keep easy to read.